plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -0,0 +1,274 @@
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+	const gchar *subsystems[] = { "wmi", NULL };
+
+	data->udev = g_udev_client_new (subsystems);
+	g_signal_connect (data->udev, "uevent",
+			  G_CALLBACK (udev_uevent_cb), plugin);
+	/* initially set to true, will wait for a device_register to reset */
+	data->needs_forcepower = TRUE;
+
+	/* determines whether to run device_registered */
+	fu_plugin_thunderbolt_power_get_path (plugin);
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_object_unref (data->udev);
+	if (data->force_path)
+		g_free (data->force_path);
+}
+
+void
+fu_plugin_device_registered (FuPlugin *plugin, FuDevice *device)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+
+	/* thunderbolt plugin */
+	if (g_strcmp0 (fu_device_get_plugin (device), "thunderbolt") == 0 &&
+	    fu_plugin_thunderbolt_power_supported (plugin)) {
+		if (data->needs_forcepower)
+			data->needs_forcepower = FALSE;
+		if (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_INTERNAL)) {
+			fu_device_set_metadata_boolean (device,
+							FU_DEVICE_METADATA_TBT_CAN_FORCE_POWER,
+							TRUE);
+		}
+	}
+}
+
+gboolean
+fu_plugin_update_prepare (FuPlugin *plugin,
+			  FuDevice *device,
+			  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GUdevDevice) udevice = NULL;
+	const gchar *devpath;
+
+	/* only run for thunderbolt plugin */
+	if (g_strcmp0 (fu_device_get_plugin (device), "thunderbolt") != 0)
+		return TRUE;
+
+	devpath = fu_device_get_metadata (device, "sysfs-path");
+
+	udevice = g_udev_client_query_by_sysfs_path (data->udev, devpath);
+	if (udevice != NULL) {
+		data->needs_forcepower = FALSE;
+		return TRUE;
+	}
+	if (!fu_plugin_thunderbolt_power_set (plugin, TRUE, error))
+		return FALSE;
+
+	data->needs_forcepower = TRUE;
+	g_usleep (TBT_NEW_DEVICE_TIMEOUT * G_USEC_PER_SEC);
+
+	return TRUE;
+}
+
+gboolean
+fu_plugin_update_cleanup (FuPlugin *plugin,
+			  FuDevice *device,
+			  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+
+	/* only run for thunderbolt plugin */
+	if (g_strcmp0 (fu_device_get_plugin (device), "thunderbolt") != 0)
+		return TRUE;
+
+	if (data->needs_forcepower &&
+	    !fu_plugin_thunderbolt_power_set (plugin, FALSE, error))
+			return FALSE;
+	return TRUE;
+}
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+
+	if (!fu_plugin_thunderbolt_power_supported (plugin)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "missing kernel support for intel-wmi-thunderbolt");
+		return FALSE;
+	}
+
+	/* this means no devices were found at coldplug by thunderbolt plugin */
+	if (data->needs_forcepower) {
+		if (!fu_plugin_thunderbolt_power_set (plugin, TRUE, error))
+			return FALSE;
+		g_timeout_add (TBT_NEW_DEVICE_TIMEOUT * 10000,
```

Comment:
No saving the return value here worries me a bit. I think it might be safer to do `data->timeout_id = g_timeout_add(` and then in destroy do `if (data->timeout_id != 0) g_source_remove (data->timeout_id)` -- this way if we unload the plugin after doing coldplug but before doing the callback we don't explode. I think you might also need to do the `g_source_remove` thing in this function too for the recoldplug stuff too.

Reply:
Fixed